File: Array/Range Sum.go
package main

import "fmt"

var mod int = 1e9 + 7

func main() {

	result := rangeSum(0, 3)
	fmt.Println(result)
}

func rangeSum(A, B int) int {
	fA := fibonacciSum(A + 1)
	fB := fibonacciSum(B + 2)

	res := fB - fA

	if res < 0 {
		res += mod
	}

	return res % mod
}

func fibonacciSum(n int) int {
	if n == 1 || n == 2 {
		return 1
	}

	var matrix [2][2]int
	matrix[0][0] = 1
	matrix[0][1] = 1
	matrix[1][0] = 1

	var matPower [2][2]int = power(matrix, n-1)

	return matPower[0][0]
}

func power(mat [2][2]int, n int) [2][2]int {
	if n == 0 {
		var matrix [2][2]int
		matrix[0][0] = 1
		matrix[0][1] = 0
		matrix[1][0] = 0
		matrix[1][0] = 1
	}

	if n == 1 {
		return mat
	}

	var half [2][2]int = power(mat, n/2)
	half = matrixMultiple(half, half)
	if n%2 == 1 {
		half = matrixMultiple(mat, half)
	}

	return half
}

func matrixMultiple(m1, m2 [2][2]int) [2][2]int {
	var res [2][2]int

	for i := 0; i < 2; i++ {
		for j := 0; j < 2; j++ {
			for k := 0; k < 2; k++ {
				res[i][j] += (m1[i][k] * m2[k][j]) % mod
				res[i][j] = res[i][j] % mod
			}
		}
	}

	return res
}

File: Pythons Solutions/Range Sum.py
mod = 1000000007

def matrixMul(m1, m2):
    res = []
    for r in range(0, 2):
        res.append([0 for c in range(0, 2)])
    
    for i in range(2):
        for j in range(2):
            for k in range(2):
                res[i][j] = (res[i][j] + m1[i][k] * m2[k][j]) % mod
                res[i][j] = res[i][j] % mod
    
    return res

def power(x, y):
    res, half = [], []
    for r in range(0, 2):
        res.append([0 for c in range(0, 2)])
        half.append([0 for c in range(0, 2)])

    if y == 0:
        res[0][0] = 1
        res[0][1] = 0
        res[1][0] = 0
        res[1][1] = 1

        return res
    if y == 1:
        return x

    half = power(x, y//2)
    half = matrixMul(half, half)
    if y % 2 == 1:
        half = matrixMul(x, half)
    
    return half


def fibonacci(n):
    if n == 1 or n == 2:
        return 1
    
    matrix, mat = [], []
    for r in range(0, 2):
        matrix.append([0 for c in range(0, 2)])
        mat.append([0 for c in range(0, 2)])

    matrix[0][0] = 1
    matrix[0][1] = 1
    matrix[1][0] = 1

    mat = power(matrix, n-1)
    return mat[0][0]

class Solution:
    def sumOfRange(self, a, b):
        fibA = fibonacci(a+1)
        fibB = fibonacci(b+2)

        res = fibB - fibA

        if res < 0:
            res += mod

        return res % mod

obj = Solution()
a, b = 0, 3
ans = obj.sumOfRange(a, b)
print(ans)